apps/emqx_management/src/emqx_mgmt_api_publish.erl

```diff
@@ -14,53 +14,80 @@
 %% limitations under the License.
 %%--------------------------------------------------------------------
 -module(emqx_mgmt_api_publish).
-%% API
+
 -include_lib("emqx/include/emqx.hrl").
+-include_lib("typerefl/include/types.hrl").
 
 -behaviour(minirest_api).
 
--import(emqx_mgmt_util, [ object_schema/1
-                        , object_schema/2
-                        , object_array_schema/1
-                        , object_array_schema/2
-                        , properties/1
-                        ]).
-
--export([api_spec/0]).
+-export([ api_spec/0
+        , paths/0
+        , schema/1
+        , fields/1
+        ]).
 
 -export([ publish/2
         , publish_batch/2]).
 
 api_spec() ->
-    {[publish_api(), publish_bulk_api()], []}.
+    emqx_dashboard_swagger:spec(?MODULE, #{check_schema => true}).
```

Comment:
#{check_schema := true, translate_body := true}

set translate_body  true, then you don't need to translate qos from binary to integer.